File: dfs.js
module.exports = (function () {

    'use strict';

    function visit(graph, fn, visited) {

        for (var node in graph) {
            if (-1 === visited.indexOf(node)) {
                visited.push(node);
                fn(node);
                visit(graph[node], fn, visited);
            }
        }
    };

    function dfs(node, graph, fn) {
        var nodes = {};
        nodes[node] = graph[node];
        visit(nodes, fn, []);
    }

    return dfs;

})();

File: golang/dfs.go
package main

func dfs(node int, nodes map[int][]int, fn func (int)) {
    dfs_recur(node, map[int]bool{}, fn)
}

func dfs_recur(node int, v map[int]bool, fn func (int)) {
    v[node] = true
    fn(node)
    for _, n := range nodes[node] {
        if _, ok := v[n]; !ok {
            dfs_recur(n, v, fn)
        }
    }
}
